distribution/xfer/download_test.go

```diff
@@ -360,3 +360,80 @@ func TestCancelledDownload(t *testing.T) {
 	close(progressChan)
 	<-progressDone
 }
+
+func TestMaxDownloadAttempts(t *testing.T) {
+	tests := []struct {
+		id                  int
+		simulateRetries     int
+		maxDownloadAttempts int
+	}{
+		// these should pass
+		{
+			id:                  1,
+			simulateRetries:     2,
+			maxDownloadAttempts: 5,
+		},
+		{
+			id:                  2,
+			simulateRetries:     5,
+			maxDownloadAttempts: 7,
+		},
+		{
+			id:                  3,
+			simulateRetries:     15,
+			maxDownloadAttempts: 1000,
+		},
+		// these should fail
+		{
+			id:                  4,
+			simulateRetries:     6,
+			maxDownloadAttempts: 5,
+		},
+		{
+			id:                  5,
+			simulateRetries:     9,
+			maxDownloadAttempts: 0,
+		},
+		{
+			id:                  6,
+			simulateRetries:     41,
+			maxDownloadAttempts: 20,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(string(tc.id), func(t *testing.T) {
+			t.Parallel()
+			layerStore := &mockLayerStore{make(map[layer.ChainID]*mockLayer)}
+			lsMap := make(map[string]layer.Store)
+			lsMap[runtime.GOOS] = layerStore
+			ldm := NewLayerDownloadManager(
+				lsMap,
+				maxDownloadConcurrency,
+				func(m *LayerDownloadManager) {
+					m.waitDuration = time.Millisecond
+					m.maxDownloadAttempts = tc.maxDownloadAttempts
+				})
+
+			progressChan := make(chan progress.Progress)
+			progressDone := make(chan struct{})
+
+			go func() {
+				for range progressChan {
+				}
+				close(progressDone)
+			}()
+
+			var currentDownloads int32
+			descriptors := downloadDescriptors(&currentDownloads)
+			descriptors[4].(*mockDownloadDescriptor).simulateRetries = tc.simulateRetries
+
+			_, _, err := ldm.Download(context.Background(), *image.NewRootFS(), runtime.GOOS, descriptors, progress.ChanOutput(progressChan))
+			if tc.id <= 3 && err != nil {
```

Comment:
I assume ID is just to help with figuring out which test failed?  Perhaps you should have a boolean for expect_failure instead of overloading ID.  It would me the logic cleaner as well.

Reply:
Yes, the ID is used to determine which test failed. In my opinion it makes sense to give feedback in which situation the test failed, but lets see what @thaJeztah has to say. Your solution would work just as fine.

Reply:
Having an ID to see which test failed is ok, and makes sense.  I meant instead not overloading ID, but it is minor.